### PR TITLE
#1039 Use merged and validate version `.rultor.yml` for `merge` command.

### PR DESCRIPTION
**Pull request by xupyprmv**
#1039 I implemented `validated` method, that checks that merged `.rultor.yml` has no changes in `architect` and `commander` sections. I added one puzzle (to implement `merged` method).

### CODE REVIEW

**Comment by alex-palevsky** [comment]
@xupyprmv Let me find a reviewer for this pull request, thanks for submitting it

**Comment by alex-palevsky** [comment]
@mkordas hlp us to review this

**Comment by mkordas** [comment]
@xupyprmv I'll do a review today

**Comment by mkordas** [comment]
@xupyprmv sorry for a delay, I'm back

**Comment by mkordas** [comment]
@xupyprmv first of all, can you ask @alex-palevsky for more time in the issue? It's been 8 days already

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 1]
```diff
@@ -83,28 +81,23 @@ public Profile fetch(final Talk talk) throws IOException {
      * Fetch a profile from an XML.
      * @param xml The XML
      * @return Profile found
+     * @todo #1039:30min Read .rultor.yml for authentication purposes from
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv I don't think it is good idea to create puzzle for that. We will have broken security before solution for this puzzle is merged

File: src/main/resources/com/rultor/agents/req/merge.sh [hunk 2]
```diff
@@ -20,6 +20,12 @@ while [ $(git show-branch "${BRANCH}" 2>/dev/null | wc -l) -gt 0 ]; do
 done
 
 git checkout -B "${BRANCH}" "fork/${fork_branch}"
+git fetch origin
+if git log --name-status --oneline HEAD..origin/${head_branch} | grep ".rultor.yml"; then
+  echo "Since original repository has different version of .rultor.yml file you have to sync your fork with the upstream repository."
+  exit 1
+fi
```
Comment (mkordas) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv I don't quite understand the above. Will it prohibit any changes to `.rultor.yml` file?

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 3]
```diff
@@ -83,28 +81,23 @@ public Profile fetch(final Talk talk) throws IOException {
...
-                            "//request/args/arg[@name='fork']/text()"
-                        ).get(0)
-                    )
-                ),
-                xml.xpath(
-                    "//request/args/arg[@name='fork_branch']/text()"
-                ).get(0)
-            );
-        }
+        final Profile profile = new GithubProfile(
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv I'm afraid we need at least some unit tests here

**Comment by mkordas** [comment]
@xupyprmv see my comments above

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 1 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv @mkordas +1

Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @original-brownbear thanks, so we share the same concern
@xupyprmv sorry, but we can't use PDD like that. After each PR & puzzle application should be production ready.

Comment (xupyprmv) [inline src/main/java/com/rultor/profiles/Profiles.java]: @mkordas Sure, you are right. Code was updated.

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (xupyprmv) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @mkordas

>  Will it prohibit any changes to .rultor.yml file?

No. You can change `.rultor.yml` as you wish. But when you try to merge, my script will check that your merge is synced with upstream. This check occurs only when `.rultor.yml` was changed in master repository. We need this to be sure that project will not crash after merge (see description #1039 for details).

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 3 above)
Comment (xupyprmv) [inline src/main/java/com/rultor/profiles/Profiles.java]: @mkordas I added unit test.

**Comment by xupyprmv** [comment]
@mkordas Great thanks for CR. I fixed PR according to your review. Please continue.

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (mkordas) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv what does it mean that merge is synced with upstream? Does it mean that it cannot be behind in terms of commits?

**Comment by mkordas** [comment]
@xupyprmv Sure, I'll continue

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import com.jcabi.xml.XMLDocument;
+import com.rultor.spi.Profile;
+import java.util.List;
+import org.apache.commons.lang3.StringUtils;
+import org.hamcrest.MatcherAssert;
+import org.junit.Ignore;
+
+/**
+ * Tests for ${@link Profiles}.
+ *
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv this empty line is unnecessary

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                String.format(profile, "Bobby", "Sly Fox", script)
+            )
+        );
+        final Profile merged = new Profiles().merged(master, fork);
+        final List<String> architects = merged.read().xpath(
+            "//entry[@key='architect']/item/text()"
+        );
+        MatcherAssert.assertThat(
+            "There is one and only one architect",
+            architects.size() == 1
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv we should use some matcher here

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        );
+        MatcherAssert.assertThat(
+            "Architect is taken from master",
+            architects.get(0).equals(architect)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv matcher would be useful here as well

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        final List<String> commanders = merged.read().xpath(
+            "//entry[@key='commander']/item/text()"
+        );
+        MatcherAssert.assertThat(
+            "There is one and only one commander",
+            commanders.size() == 1
+        );
+        MatcherAssert.assertThat(
+            "Commander is taken from master",
+            commanders.get(0).equals(commander)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv I won't be repeating myself - we need proper matchers everywhere

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 8]
```diff
@@ -73,26 +72,37 @@ public Profile fetch(final Talk talk) throws IOException {
...
             } else {
-                profile = Profiles.fetch(xml);
+                profile = this.fetch(xml);
             }
         }
         return profile;
     }
 
     /**
+     * Merge profile from master and fork.
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv can we have more info in Javadoc describing purpose and details of merging? I'm ok to add this task to todo

**Comment by mkordas** [comment]
@xupyprmv see my comments

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (xupyprmv) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @mkordas Yes, your fork cannot be behind, if syncing leads to update `.rultor.yml`. Otherwise we can have situation: master is OK, branch is OK, but after merge build is broken. For example: you make a fork, and add new script. After that master `.rultor.yml` is updated (new script). Both master and your branch is OK. But merged result fails (because this two script can be incompatible). If `.rultor.yml` is synced then this situation will not occurs.

**Comment by xupyprmv** [comment]
@mkordas I updated PR. Please continue with CR.

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (mkordas) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv makes a lot of sense, thanks for adding it. I think ideally we should have a test for that though... In this case puzzle should be fine I believe, but it will be up to @original-brownbear

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        );
+        MatcherAssert.assertThat(
+            "There is one and only one script",
+            scripts.size(),
+            Matchers.is(1)
+        );
+        MatcherAssert.assertThat(
+            "Script is taken from fork",
+            scripts.get(0),
+            Matchers.is(script)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv there is `contains` matcher that verifies both size and contents at once

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+             architects.get(0),
+             Matchers.is(architect)
+        );
+        final List<String> commanders = merged.read().xpath(
+            "//entry[@key='commander']/item/text()"
+        );
+        MatcherAssert.assertThat(
+            "There is one and only one commander",
+            commanders.size(),
+            Matchers.is(1)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv if we used `contains` mather then this check would become obsolete

**Comment by mkordas** [comment]
@xupyprmv so there's just matcher and test thing left, all else is good

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (original-brownbear) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv @mkordas to me this check seems a little redundant to be honest, in fact it might even be a bit of a usability problem for users not used to consistently rebasing. But feel free to correct me if I'm wrong here :)

I mean we are not Travis. The whole point of this task is to run the merge result `rultor.yml`, not the one from master or the fork individually (**no individually is very important**). 
The `rultor.yml` executed must be exactly that, which will end up in the repo after the merge.
This is the only way to ensure that past issues with the merge result introducing some dependency issue from the interplay `rultor.yml` and source code will not happen ever again.
What you build is what you get so to say ;)

Comment (xupyprmv) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @original-brownbear @mkordas Yes, the best way will be use merged version of `.rultor.yml`. But there is a deadlock: to merge `.rultor.yml` we need to have merged `.rultor.yml` file. This task is not an easy one to solve it inside rultor (in case, if we are not constructing AI here). I think it is better to move merging (and solving possible conflicts) to PR author side. Yes it can be noisy enough (if `.rultor.yml` is changed very frequently) but I don't see any better solution here.

**Comment by xupyprmv** [comment]
@mkordas Thanks. I updated my PR. `Matchers` looks much better now.

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (original-brownbear) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv no don't overthink this :) No AI needed here. Otherwise I wouldn't have stated this so clearly in the issue description. Using the merge result here is really without alternative, sorry :)

Getting the merge result is actually fairly easy and can be accomplished in two different ways I can think of right now:

* Fetch the merge result from GitHub, see [here](https://help.github.com/articles/checking-out-pull-requests-locally/#modifying-an-inactive-pull-request-locally) 
* **Merge it on the cli side and use the result in Rultor (just cat the result and use it as Shell output)** (This is the one we need to implement, it's 100% safe, because this is what Rultor actually is going to push eventually.)
 
Rultor does not solve any conflicts btw, this needs to be done by the author. All we really need to ensure is that we are actually building against the resulting `.rultor.yml`. This is an easy thing to x

Comment (xupyprmv) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @original-brownbear 

> Getting the merge result is actually fairly easy and can be accomplished in two different ways I can think of right now

Problem is not to merge forks. It is easy task. Problem is to be sure, that we have correct version of `.rultor.yml` here: https://github.com/yegor256/rultor/pull/1064/files#diff-eb31f19e43e00bbef161a0fbd2a228a0R93 . And in current architecture of Rultor you cannot merge `.rultor.yml` before creating profile. So, the only way here - sync fork with upstream.

> Rultor does not solve any conflicts btw, this needs to be done by the author. All we really need to ensure is that we are actually building against the resulting .rultor.yml.

Yes, this is what I speaking about. Let fork author to do it himself.

Comment (original-brownbear) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @xupyprmv I know that the current architecture cannot properly do this, we need to make a change obviously :) 
The problem with requiring `sync fork with upstream` is that this will inevitably cause builds that worked fine before to fail. We cannot add a requirement to sync your `.rultor.yml` with `master` to Rultor, just because it's harder to implement this properly on the Rultor end.

Also, I mean we're both clear on how to do this now I think. We get the initial profile from master, use it to answer the GitHub request.
Then get the merged profile and use it for the build itself, plus check that it doesn't adjust  information about who's commander for what and such.

So basically the three things we need are:
* some functionality to fetch the merged profile.
* some functionality to verify it does not contain changes to authentication things
* adjust the actual build to run on the merged profile

=> my suggestion to adjust the solution here would be this:
* create puzzles for the above three things where they are missing
* remove changes to the bash script
* Fix whatever @mkordas finds

=> we're good :)

**Comment by mkordas** [comment]
@xupyprmv looks all right, thanks!

**Comment by mkordas** [comment]
@rultor merge

File: src/main/resources/com/rultor/agents/req/merge.sh (hunk 2 above)
Comment (xupyprmv) [inline src/main/resources/com/rultor/agents/req/merge.sh]: @original-brownbear Great thanks for advices. I'm really appreciate your help.

> some functionality to fetch the merged profile.

I added puzzle.

> some functionality to verify it does not contain changes to authentication things

I implemented method and covered it by unit tests.

> adjust the actual build to run on the merged profile

We will have this out-of-the box here:  https://github.com/yegor256/rultor/pull/1064/files#diff-eb31f19e43e00bbef161a0fbd2a228a0R152

**Comment by xupyprmv** [comment]
@mkordas Please review new version. I totally refactored code according to @original-brownbear remarks.

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 11]
```diff
@@ -73,18 +79,66 @@ public Profile fetch(final Talk talk) throws IOException {
...
+     * @return Validated profile
+     * @throws IOException If fails
+     */
+    public Profile validated(final Profile master, final Profile merged)
+        throws IOException {
+        final String[] security = {"architect", "commander"};
+        for (final String section : security) {
+            if (!Arrays.equals(
+                master.read().xpath(
+                    String.format(Profiles.ITEMS, section)
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv there's some code duplication to read, get xpath and get array - maybe we could extract it to method?

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        "<entry key='commander'><item>%s</item></entry>",
+        "<entry key='script'><item>%s</item></entry>",
+        "</entry></p>"
+    );
+
+    /**
+     * Profile validation fails if architect list was changed.
+     * @throws Exception In case of error.
+     */
+    @Test(expected = Profile.ConfigException.class)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv can we use try-catch here and perform assertion on exception message?

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            )
+        );
+        new Profiles().validated(master, merged);
+    }
+
+    /**
+     * Merged profile validation fails if architect list was changed.
+     * @throws Exception In case of error.
+     */
+    @Test(expected = Profile.ConfigException.class)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv same here

**Comment by mkordas** [comment]
@xupyprmv see my comments

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 11 above)
Comment (xupyprmv) [inline src/main/java/com/rultor/profiles/Profiles.java]: @mkordas I extracted method.

File: src/test/java/com/rultor/profiles/ProfilesTest.java (hunk 12 above)
Comment (xupyprmv) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @mkordas Sure. Fixed. Can you share (if possible) some info why approach with try-cathch is better then  using`expcected`?

**Comment by xupyprmv** [comment]
@mkordas Thanks for CR. I updater PR. Please continue.

File: src/test/java/com/rultor/profiles/ProfilesTest.java (hunk 12 above)
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv try-catch enables to validate exception message/cause and also is more precise where exception is allowed to happen and where not

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+                String.format(
+                    ProfilesTest.PROFILE,
+                    architect,
+                    "Bob",
+                    "do_another2"
+                )
+            )
+        );
+        try {
+            new Profiles().validated(master, merged);
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv we need `Assert.fail` here

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+                    ProfilesTest.PROFILE,
+                    "Bobby",
+                    commander,
+                    "do_another"
+                )
+            )
+        );
+        try {
+            new Profiles().validated(master, merged);
+            MatcherAssert.assertThat(
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv wouldn't it be better to use `Assert.fail` instead of assertion on `false`?

**Comment by xupyprmv** [comment]
@mkordas PR was fixed according your remarks. Please continue.

**Comment by mkordas** [comment]
@xupyprmv looks all right, thanks!

**Comment by mkordas** [comment]
@rultor merge

**Comment by original-brownbear** [comment]
@xupyprmv I'm on it :)

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 16]
```diff
@@ -73,18 +79,77 @@ public Profile fetch(final Talk talk) throws IOException {
...
+     * from master branch.
+     * @param master Profile from master branch
+     * @param fork Profile from fork branch
+     * @return Merged profile
+     * @todo #1039:30min Modify method implementation: merged profile is a
+     *  `.rultor.yml` form result of real `git merge` of master branch and
+     *  fork branch. We need real `git merge` result here. See some details
+     *  here: https://github.com/yegor256/rultor/pull/1064/files#r56796959
+     */
+    public Profile merged(final Profile master, final Profile fork) {
```
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv This is the only slight issue in this PR that I can find :)
Why do we need the fork profile here ? :)

In my opinion we don't because as you explain just fine in the Puzzle, the actual merge is going to happen on the EC2 instance using actual GIT :) That thing only needs to know the name of the fork (the coordinates so to say ;)).
I suggest simply making the second input the coordinate string here. Then we can pass that to GIT => lots easier :)
I'll explain one more detail on this below ;) Here really just make the second input `final String fork`.

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 17]
```diff
@@ -92,17 +157,26 @@ private static Profile fetch(final XML xml) {
...
+            final Profile master = new GithubProfile(
+                    new TalkIssues(Profiles.github(), xml).get().repo()
+            );
+            profile = this.validated(
+                master,
+                this.merged(
+                    master,
+                    new GithubProfile(
+                        Profiles.github().repos().get(
+                            new Coordinates.Simple(
```
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv just pass this String here:
```
                               xml.xpath(
                                    "//request/args/arg[@name='fork']/text()"
                                ).get(0)
```

 instead of the actual fetched repo ;) instantly saves us one call to the GitHub API here, Rultor will eventually have to fetch the fork contents anyhow, no need to know those here already imo :)

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 18]
```diff
@@ -73,18 +79,77 @@ public Profile fetch(final Talk talk) throws IOException {
...
+     * Validate merged profile: merged profile must have no changes in architect
+     * and commander sections (full match with master profile).
+     * @param master Profile from master branch
+     * @param merged Result of merging master and fork profiles
+     * @return Validated profile
+     * @throws IOException If fails
+     */
+    public Profile validated(final Profile master, final Profile merged)
+        throws IOException {
+        final String[] security = {"architect", "commander"};
```
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv the correct key here is `commanders`. Also here we need to make sure that we check this for all commands nested. If you check the `.rultor.yml`, you'll find that we have `commanders` for every command.

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 19]
```diff
@@ -73,18 +79,77 @@ public Profile fetch(final Talk talk) throws IOException {
...
+        for (final String section : security) {
+            if (!Arrays.equals(
+                section(master, section),
+                section(merged, section)
+            )) {
+                throw new Profile.ConfigException(
```
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv are you sure this will eventually be posted as an Answer in GitHub ?

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+ */
+public class ProfilesTest {
+
+    /**
+     * Profile template.
+     */
+    private static final String PROFILE = StringUtils.join(
+        "<p><entry key='merge'>",
+        "<entry key='architect'><item>%s</item></entry>",
+        "<entry key='commander'><item>%s</item></entry>",
```
Comment (original-brownbear) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv `commanders` again + this is nested within the three other elements `merge`, `deploy`, `release`.

**Comment by original-brownbear** [comment]
@xupyprmv nice work :) Just two relatively easy things overall, check my comments.

With the merge result, you just need to delete a few lines and make a minor adjustment to the function => 2 minutes :)

The profile validation is a little iffy, I think you simply got the `commanders` key a little wrong. Commanders really are defined per command, there is no overall `commander` :)
The general approach here looks nice though. You have two choices here:

1. Drop everything commander related and just keep the already correct `architect` part of the validation + add a puzzle.
2. Just correct the `commanders` section validation and make it check the subkeys under `merge`, `deploy` and `release`.

Choice is yours ;) Of course I prefer 2 and for me it looks like this is an easy adjustment for you to make from what you already have. But feel free to add the puzzle with a proper explanation.

**Comment by mkordas** [comment]
@xupyprmv do you need help from me here?

**Comment by xupyprmv** [comment]
@mkordas Can you review new version? I added some modifications according to @original-brownbear remarks.

**Comment by mkordas** [comment]
@xupyprmv sure, I'm on it

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 21]
```diff
@@ -73,18 +73,93 @@ public Profile fetch(final Talk talk) throws IOException {
...
+     * @todo #1039:30min Modify method implementation: merged profile is a
+     *  `.rultor.yml` form result of real `git merge` of master branch and
+     *  fork branch. We need real `git merge` result here. See some details
+     *  here: https://github.com/yegor256/rultor/pull/1064/files#r56796959
+     */
+    public Profile merged(
+        final Profile master,
+        final String fork,
+        final String branch
+    ) {
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv we do not use paired brackets notation for method parameters, I'd put here as much as possible into each line

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 19 above)
Comment (xupyprmv) [inline src/main/java/com/rultor/profiles/Profiles.java]: @original-brownbear I'm not sure. I think this exception will be maintained here: 

https://github.com/xupyprmv/rultor/blob/%231039/src/main/java/com/rultor/Routine.java#L129

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 22]
```diff
@@ -73,18 +73,93 @@ public Profile fetch(final Talk talk) throws IOException {
...
+    }
+
+    /**
+     * Validate merged profile: merged profile must have no changes in architect
+     * and commander sections (full match with master profile).
+     * @param master Profile from master branch
+     * @param merged Result of merging master and fork profiles
+     * @return Validated profile
+     * @throws IOException If fails
+     * @checkstyle MultipleStringLiteralsCheck (10 lines)
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv if it must be the same string then we should rather extract variable

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 23]
```diff
@@ -73,18 +73,93 @@ public Profile fetch(final Talk talk) throws IOException {
...
+    }
+
+    /**
+     * Get section content.
+     * @param profile Profile
+     * @param section Section of profile
+     * @return Content of section as array of strings
+     * @throws IOException If fails
+     */
+    private String[] section(final Profile profile, final String... section)
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv shouldn't it be `sections` now?

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 24]
```diff
@@ -73,18 +73,93 @@ public Profile fetch(final Talk talk) throws IOException {
...
+        throws IOException {
+        final StringBuilder path = new StringBuilder(Tv.HUNDRED);
+        for (final String element : section) {
+            path.append(String.format("/entry[@key='%s']", element));
+        }
+        path.append("");
```
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv why we are appending empty string?

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * Tests for ${@link Profiles}.
+ * @author Maksimenko Vladimir (xupypr@xupypr.com)
+ * @version $Id$
+ * @since 1.62
+ * @checkstyle MultipleStringLiteralsCheck (500 lines)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv this check is valuable for tests, as each test should use it's own data. Can we somehow avoid suppressing it or at least make it more local?

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+    );
+
+    /**
+     * Template for exception message.
+     */
+    private static final String MESSAGE =
+        "You cannot change `%s` section for security reasons";
+
+    /**
+     * Profile validation fails if architect list was changed.
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv second word in test should be `can`

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 27]
```diff
@@ -0,0 +1,311 @@
...
+                        ProfilesTest.MESSAGE,
+                        "[architect]"
+                    )
+                )
+            );
+        }
+    }
+
+    /**
+     * Merged profile validation fails if architect list was changed.
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv we should have `can` here as well

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 28]
```diff
@@ -0,0 +1,311 @@
...
+     * @throws Exception In case of error.
+     * @checkstyle MultipleStringLiteralsCheck (30 lines)
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv this is kind of duplicate suppression now

File: src/test/java/com/rultor/profiles/ProfilesTest.java [hunk 29]
```diff
@@ -0,0 +1,311 @@
...
+                        ProfilesTest.MESSAGE,
+                        "[merge, commanders]"
+                    )
+                )
+            );
+        }
+    }
+
+    /**
+     * Merged profile validation fails if architect list was changed.
```
Comment (mkordas) [inline src/test/java/com/rultor/profiles/ProfilesTest.java]: @xupyprmv remember about `can`

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 23 above)
Comment (xupyprmv) [inline src/main/java/com/rultor/profiles/Profiles.java]: @mkordas No. Because String array is a section, described as sequence of nested entries.

**Comment by xupyprmv** [comment]
@mkordas Thanks. I fixed my PR according to your remarks. Please continue.

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 23 above)
Comment (mkordas) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv OK, I see

**Comment by mkordas** [comment]
@xupyprmv checking again

**Comment by mkordas** [comment]
@xupyprmv looks OK

**Comment by mkordas** [comment]
@rultor merge

File: src/main/java/com/rultor/profiles/Profiles.java [hunk 30]
```diff
@@ -73,36 +78,108 @@ public Profile fetch(final Talk talk) throws IOException {
...
+        final String commanders = "commanders";
+        final String[][] security = {
+            {"architect"},
+            {Profiles.MERGE, commanders},
+            {"deploy", commanders},
+            {"release", commanders},
+        };
+        for (final String[] section : security) {
+            if (!Arrays.equals(
+                section(master, section),
```
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv  shouldn't it be `this.section()` here for the method call ? ( we never leave out the `this` :) )

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 23 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv @mkordas what am I missing here :) ? Shouldn't this be `static` ?

File: src/main/java/com/rultor/profiles/Profiles.java (hunk 19 above)
Comment (original-brownbear) [inline src/main/java/com/rultor/profiles/Profiles.java]: @xupyprmv oh no that's something else, that's the thing that generates this green dotted timeline on http://rultor.com :)

I suggest to simply remove the call to this function for now then, add to the puzzle that this should be used on the actual merge result eventually ( it seems to work nicely looking at the tests! :) ), but don't use it as of now in production code. Not worth adding a random risk here, it's simply not properly tested at this point and a little tricky to test too, you've done enough here no need for you to add the IT case for that. Just adjust puzzle and remove the call to this for now :)
Sounds reasonable ?

**Comment by xupyprmv** [comment]
@original-brownbear Thanks for review. Please look at new version. I modified code and added one more puzzle.

**Comment by original-brownbear** [comment]
@xupyprmv good job! :), but please do me one last  favour and merge both puzzles into one :)
Really just ask for both things IT + merged profile to be done in one puzzle.

It's just a mistake I made here before to have parallel running puzzles that should better be done consecutively, this is exactly this kind of case ;)

=> just then puzzle and I'll hit the merge :)

**Comment by xupyprmv** [comment]
@original-brownbear I merged puzzles.

**Comment by original-brownbear** [comment]
@xupyprmv nice thanks!

**Comment by original-brownbear** [comment]
@rultor merge please :)

**Comment by original-brownbear** [comment]
@rultor merge please :)

**Comment by original-brownbear** [comment]
@xupyprmv sorry not gonna work this way you need to squash your commits into one ;)

Preferably just do this:

```
git fetch origin
git rebase --interactive master
```

then just fixup all commits except the last one and reword the last one accordingly ;)

**Comment by original-brownbear** [comment]
@xupyprmv don't close this one please, just force push on it :) Otherwise this turns payment for CR weird ;)

**Comment by xupyprmv** [comment]
@original-brownbear sorry, it was automatic action of my git tool.  I will try to fix it.

**Comment by original-brownbear** [comment]
@xupyprmv  np please reopen asap though ;)

**Comment by xupyprmv** [comment]
@original-brownbear Sorry, looks like github prevents me reopen PR. https://github.com/isaacs/github/issues/361

**Comment by xupyprmv** [comment]
@original-brownbear I'm really sorry for this. I will keep in mind not to do such actions in future. But there is no way to reopen this PR. Can I create new PR?

**Comment by original-brownbear** [comment]
@xupyprmv sure I'll handle that new one asap then :)

**Comment by mkordas** [comment]
@xupyprmv yeah, unfortunately GitHub doesn't handle force pushes to closed PRs :(

**Comment by xupyprmv** [comment]
@mkordas Sorry for this. I will avoid this situation next time.

**Comment by mkordas** [comment]
@xupyprmv no worries, I think it's rather GitHub's fault in this case :) Thanks for your huge effort - issue was pretty hard.

**Comment by alex-palevsky** [comment]
@mkordas 10 mins added to @original-brownbear account (our architect), payment ID is `AP-0SD80845N0803935L`, since quality is good here

Thank you, I have added **1 hour and 3 mins** to your account in payment/transaction "`56f93cbb7c7beb01b600005d`", time consumed: 167 hours and 34 mins

extra minutes for review comments (c=48)

+63 added to your rating, at the moment it is: [+9124](http://www.netbout.com/b/36080?open=rating)
